File: rosbasic/src/actionServer.cpp
#include <ros/ros.h>
#include <std_msgs/Int32.h>
#include <actionlib/server/simple_action_server.h>
#include "rosbasic/CounterAction.h"
#include <iostream>
#include <sstream>

class GhostMan{
protected:
	ros::NodeHandle nh;
	actionlib::SimpleActionServer<rosbasic::CounterAction> as;

	rosbasic::CounterFeedback feedback;
	rosbasic::CounterResult result;

	std::string action_name;
	int goal;
	int progress;
public:
	GhostMan(std::string name) : as(
		nh,
		name,
		boost::bind(&GhostMan::executeCB,this,_1),
		false),
	action_name(name)
	{

		as.registerPreemptCallback(boost::bind(&GhostMan::preemptCB,this));
		as.start();
	}

	void preemptCB()
	{
		ROS_WARN("%s got preempted!", action_name.c_str());
		result.final_count=progress;
		as.setPreempted(result,"I got Preempted");
	}
	void executeCB(const rosbasic::CounterGoalConstPtr &goal)
	{
		if(!as.isActive() || as.isPreemptRequested()) return;

		ros::Rate r(5);
		ROS_INFO("%s is processing the goal %d",action_name.c_str(),goal->count);

		for(progress =1;progress<=goal->count;progress++)
		{
			if(!ros::ok())
			{
				result.final_count = progress;
				as.setAborted(result,"I fail");
				ROS_INFO("%s shutting dodwn",action_name.c_str());
				break;
			}
			if(!as.isActive() || as.isPreemptRequested() ) return;

			if(goal->count<=progress)
			{
				result.final_count = progress;
				as.setSucceeded(result);

			}
			else{
				feedback.current_number = progress;
				as.publishFeedback(feedback);
			}
			r.sleep();
		}
	}// end exeCB

};


int main(int argc,char** argv)
{
	ros::init(argc,argv,"action_server");
	ROS_INFO("Starting Action server");
	//GhostMan gm(ros::this_node::getName());
	GhostMan gm("Counter_action");
	ros::spin();
	return 0;
}

File: rosbasic/script/action_server.py
#! /usr/bin/env python

import actionlib
import rospy

from rosbasic.msg import *

class CounterClass(object):

	_feedback = CounterFeedback()
	_result   = CounterResult()

	_feedback.current_number = 0

	def __init__(self):
		self._as = actionlib.SimpleActionServer("/counter", CounterAction, self.goal_callback, False)
		self._as.start()

	def goal_callback(self, goal):
		r = rospy.Rate(1)
		success = True

		progress = goal.count
		for i in range(0, progress):

			if self._as.is_preempt_requested():
				rospy.loginfo("The goal has been cancelled")
				self._as.set_preempted()
				success = False
				break

			self._feedback.current_number += 1
			self._as.publish_feedback(self._feedback)

			r.sleep()

		if success:
			self._result.final_count = self._feedback.current_number
			rospy.loginfo("Succeeded! Final goal, count = %s" % self._result.final_count)
			self._as.set_succeeded(self._result)

if __name__ == '__main__':
	rospy.init_node("action_server_node")
	CounterClass()
	rospy.spin()